src/VisualStudio/Next/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs

```diff
@@ -0,0 +1,178 @@
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Execution;
+using Microsoft.CodeAnalysis.Extensions;
+using Microsoft.VisualStudio.LanguageServices.Implementation.Remote;
+
+namespace Microsoft.VisualStudio.LanguageServices.Remote
+{
+    internal partial class RemoteHostClientServiceFactory
+    {
+        private class RemoteHostClientService : IRemoteHostClientService
+        {
+            private readonly Workspace _workspace;
+            private readonly IDiagnosticAnalyzerService _analyzerService;
+
+            private readonly object _gate;
+
+            private CancellationTokenSource _shutdown;
+            private Task<RemoteHostClient> _instance;
+
+            public RemoteHostClientService(Workspace workspace, IDiagnosticAnalyzerService analyzerService)
+            {
+                _gate = new object();
+
+                _workspace = workspace;
+                _analyzerService = analyzerService;
+            }
+
+            public void Enable()
+            {
+                lock (_gate)
+                {
+                    if (_instance != null)
+                    {
+                        // already enabled
+                        return;
+                    }
+
+                    if (!_workspace.Options.GetOption(RemoteHostOptions.RemoteHost))
+                    {
+                        // not turned on
+                        return;
+                    }
+
+                    // make sure we run it on background thread
+                    _shutdown = new CancellationTokenSource();
+
+                    var token = _shutdown.Token;
+                    _instance = Task.Run(() => EnableAsync(token), token);
+                }
+            }
+
+            public void Disable()
+            {
+                lock (_gate)
+                {
+                    if (_instance == null)
+                    {
+                        // already disabled
+                        return;
+                    }
+
+                    var instance = _instance;
+                    _instance = null;
+
+                    RemoveGlobalAssets();
+
+                    _shutdown.Cancel();
+
+                    try
+                    {
+                        instance.Result.Shutdown();
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        // _instance wasn't finished running yet.
+                    }
+                }
+            }
+
+            public async Task<RemoteHostClient> GetRemoteHostClientAsync(CancellationToken cancellationToken)
+            {
+                // copy instance to local variable so that we don't need lock here
+                var instance = _instance;
```

Comment:
You still need lock.  Otherwise reads/writes can be reordered.  For example, the _instance field may be in the process of being partially constructed even as you read it.

Please don't put in these sorts of micro-optimizations as they can lead to very subtle bugs.